### DISCUSSION TITLE
how libvips handles the floating number deep down?

### DISCUSSION THREAD

**Discussion by gemini133** [libvips/libvips#3932] [General]
image: 300x271.

scale factor = 0.99999999999998 (14 decimals)

with resize method
```
newImage = image.resize(factor)
```
newImage: 300x271

with thumbnail_image method

```
newWidth = mul(300, factor, 14) // 299.99999999999397
newHeight = mul(271, factor, 14) // 270.99999999999460
newImage = image.thumbnail_image(newWidth, { height: newHeight, size: 'force' })
```
newImage : 299x270

**Comment by jcupitt** [comment]
Hi @gemini133,

`resize` uses round to nearest for the image dimensions. For example:

```
$ vipsheader k2.jpg
k2.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
```

1450 / 1450.5 is 0.99965, so:

```
$ vips resize k2.jpg x.jpg 0.9997
$ vipsheader x.jpg
x.jpg: 1450x2047 uchar, 3 bands, srgb, jpegload
```

That's just over 1449.5, so it rounds up to 1450. But:

```
$ vips resize k2.jpg x.jpg 0.9995
$ vipsheader x.jpg
x.jpg: 1449x2047 uchar, 3 bands, srgb, jpegload
```

Now it's just below and it rounds down.

In your example code:

```
newWidth = mul(300, factor, 14) // 299.99999999999397
newHeight = mul(271, factor, 14) // 270.99999999999460
newImage = image.thumbnail_image(newWidth, { height: newHeight, size: 'force' })
```

`newWidth` and `newHeight` are passed as ints, so they will become 299 and 270 before libvips sees them. You'll need to use something like `rint()` to get the same behaviour as resize.

You can see parameter types like this:

```
$ vips thumbnail
generate thumbnail from file
usage:
   thumbnail filename out width [--option-name option-value ...]
where:
   filename     - Filename to read from, input gchararray
   out          - Output image, output VipsImage
   width        - Size to this width, input gint
			default: 1
			min: 1, max: 100000000
optional arguments:
   height       - Size to this height, input gint
			default: 1
			min: 1, max: 100000000
...
```

Please don't use `thumbnail_image` -- it's only there for emergencies. Plain `thumbnail` will often be many, many times faster and use a lot less memory (I expect this is just example code and you're not doing this in your application).

**Comment by gemini133** [comment]
lol, my mistake
